src/object/authenticate_access_auth.cpp

```diff
@@ -1136,12 +1175,691 @@ au_user_revoke_all_privileges (MOP user_mop)
+ *   obj_type(in): the object type
+ *   old_owner_mop(in): class/stored procedure old owner
+ *   new_owner_mop(in): class/stored procedure new owner
+ *   unique_name(in):
+ * NOTE
+ * When the owner of a class, virtual class, or procedure is changed, the previous owner's privileges are transferred to the new owner.
+ *
+ * However, if the new owner already possesses the privileges granted by the previous owner, those privileges are removed.
+ * Reason: The REVOKE statement cannot revoke privileges from the owner.
+ */
+int
+au_object_owner_change_privileges (DB_OBJECT_TYPE obj_type, MOP object_mop, MOP old_owner_mop, MOP new_owner_mop,
+				   const char *unique_name)
+{
+  int error = NO_ERROR;
+  int update_count_db_authorization = 0;
+
+  assert (old_owner_mop != NULL && new_owner_mop != NULL && unique_name != NULL);
+
+  /* modify db_authorization catalog */
+  error = update_authorization_for_new_owner (obj_type, old_owner_mop, new_owner_mop, unique_name,
+	  &update_count_db_authorization);
+  if (error != NO_ERROR)
+    {
+      ASSERT_ERROR_AND_SET (error);
+      goto exit;
+    }
+
+  /* if there are no results from querying the db authorization catalog, there is no need to check db_auth. */
+  if (update_count_db_authorization)
+    {
+      /* modify db_auth catalog */
+      error = update_auth_for_new_owner (obj_type, old_owner_mop, new_owner_mop, unique_name);
+      if (error != NO_ERROR)
+	{
+	  ASSERT_ERROR_AND_SET (error);
+	  goto exit;
+	}
+
+      if (obj_type == DB_OBJECT_CLASS)
+	{
+	  SM_CLASS *classobj;
+	  if ((error = au_fetch_class_force (object_mop, &classobj, AU_FETCH_READ)) == NO_ERROR)
+	    {
+	      /*
+	       * clear the cache for this user/class pair to make sure we
+	       * recalculate it the next time it is referenced
+	       */
+	      Au_cache.reset_cache_for_user_and_class (classobj);
+	    }
+	}
+
+      /*
+       * Make sure that we don't keep any parse trees
+       * around that rely on obsolete authorization.
+       * This may not be necessary.
+       */
+      sm_bump_local_schema_version ();
+    }
+
+exit:
+  return (error);
+}
+
+static int
+update_authorization_for_new_owner (DB_OBJECT_TYPE obj_type, MOP old_owner_mop, MOP new_owner_mop,
+				    const char *unique_name, int *row_count)
+{
+  int error = NO_ERROR, save, current_cache;
+  char obj_fetch_query[256];
+  const char *sql_query =
+	  "SELECT [au].grantee, [au].object_of FROM [" CT_CLASSAUTH_NAME "] [au]"
+	  " WHERE [au].[object_of] = (%s)"
+	  " GROUP BY [au].grantee";
+  DB_VALUE val, element;
+  DB_SESSION *session = NULL;
+  int stmt_id;
+  DB_QUERY_RESULT *result = NULL;
+  DB_VALUE grantee_value, object_of_value;
+  MOP grantor_mop = NULL, grantee_mop = NULL, object_of_mop = NULL, auth = NULL;
+  DB_SET *grants = NULL;
+  int gindex, gsize;
+  std::unordered_map<std::tuple<MOP, MOP, MOP>, int,
+      authorization_keyhash, authorization_keyequal> authorization_unordered_map;
+  std::tuple<MOP, MOP, MOP> key;
```

Comment:
```suggestion
  std::unordered_map<AuthorizationKey, int,
      tuple_hash<AuthorizationKey>, tuple_equal<AuthorizationKey>> authorization_unordered_map;
  AuthorizationKey key;
```
첫 코맨트를 적용한다면 다음과 같이 사용이 가능합니다.

Reply:
코드 리뷰 감사합니다.
리뷰 내용을 반영하여 수정했습니다.